File: legacy-python/work_agent/execution/pr_ops.py
from __future__ import annotations

from typing import Protocol

from . import shell
from .models import RepoInfo


class PullRequestCreator(Protocol):
    """No implementation of this may ever merge a PR — creation only, and
    always as a draft."""

    def create_draft_pr(self, repo: RepoInfo, branch: str, title: str, body: str) -> str:
        """Returns the PR URL."""
        ...


class GhCliPullRequestCreator:
    """Uses the `gh` CLI (already authenticated with real org access in this
    environment — the GitHub MCP plugin's PAT is scoped to personal repos
    only and can't see org repos, so this is the mechanism that actually
    works here). Always passes --draft; never calls `gh pr merge`."""

    def create_draft_pr(self, repo: RepoInfo, branch: str, title: str, body: str) -> str:
        cmd = [
            "gh", "pr", "create",
            "--repo", repo.full_name,
            "--head", branch,
            "--base", repo.default_branch,
            "--draft",
            "--title", title,
            "--body", body,
        ]
        result = shell.run_or_raise(cmd, cwd=repo.local_path, timeout=120)
        lines = [line.strip() for line in result.output.strip().splitlines() if line.strip()]
        return lines[-1] if lines else ""


class MockPullRequestCreator:
    def __init__(self):
        self.calls: list[dict] = []

    def create_draft_pr(self, repo: RepoInfo, branch: str, title: str, body: str) -> str:
        self.calls.append({"repo": repo.full_name, "branch": branch, "title": title, "body": body})
        return f"https://github.com/{repo.full_name}/pull/999"

File: src/agents/engineeringExecution/prOps.ts
import { runOrThrow } from './shell.js';
import type { RepoInfo } from './models.js';
import { repoFullName } from './models.js';

/** No implementation of this may ever merge a PR — creation only, draft or
 * not. `asDraft` is caller-supplied (from config.engineeringExecution.
 * openPrAsDraft, true by default) — this interface itself takes no
 * position on which is safer, that decision lives in config. */
export interface PullRequestCreator {
  createPr(repo: RepoInfo, branch: string, title: string, body: string, asDraft: boolean): Promise<string>;
}

/** Uses the `gh` CLI (already authenticated with real org access in this
 * environment — the GitHub MCP plugin's PAT is scoped to personal repos
 * only, confirmed during Phase 2's first live run). Never calls `gh pr
 * merge` — that guardrail is unconditional, regardless of `asDraft`. */
export class GhCliPullRequestCreator implements PullRequestCreator {
  async createPr(repo: RepoInfo, branch: string, title: string, body: string, asDraft: boolean): Promise<string> {
    const cmd = [
      'gh', 'pr', 'create',
      '--repo', repoFullName(repo),
      '--head', branch,
      '--base', repo.defaultBranch,
      ...(asDraft ? ['--draft'] : []),
      '--title', title,
      '--body', body,
    ];
    const result = await runOrThrow(cmd, { cwd: repo.localPath, timeoutMs: 120_000 });
    const lines = result.output.trim().split('\n').map((l) => l.trim()).filter(Boolean);
    return lines.at(-1) ?? '';
  }
}

export class MockPullRequestCreator implements PullRequestCreator {
  calls: Array<{ repo: string; branch: string; title: string; body: string; asDraft: boolean }> = [];

  async createPr(repo: RepoInfo, branch: string, title: string, body: string, asDraft: boolean): Promise<string> {
    this.calls.push({ repo: repoFullName(repo), branch, title, body, asDraft });
    return `https://github.com/${repoFullName(repo)}/pull/999`;
  }
}
